thetis/callback.py

```diff
@@ -5,35 +5,159 @@
 """
 from __future__ import absolute_import
 from .utility import *
+from abc import ABCMeta, abstractproperty, abstractmethod
+import h5py
 
 
-class ValidationCallback(object):
+class CallbackManager(object):
     """
-    Base class of all objects used to check validity of simulation
-    at each export.
+    Stores callbacks in different categories and provides methods for
+    evaluating them.
     """
-    def __init__(self):
-        self._initialized = False
+    def __init__(self, modes=('export', 'timestep')):
+        self.callbacks = {}
+        for mode in modes:
+            self.callbacks[mode] = OrderedDict()
 
-    def initialize(self, comm=COMM_WORLD):
-        """Initializes the callback function.
+    def add(self, callback, mode):
+        key = callback.name
+        self.callbacks[mode][key] = callback
 
-        :kwarg comm: The communicator for the callback."""
-        self._initialized = True
+    def evaluate(self, solver_obj, mode):
+        for key in self.callbacks[mode]:
+            self.callbacks[mode][key].evaluate(solver_obj)
```

Comment:
Also, I would then inherit from `defaultdict`:

```python
class CallbackManager(defaultdict):
    def __init__(self):
        super(CallbackManager, self).__init__(OrderedDict)
```

That way you don't need to implement `__getitem__`.